ts/test/session/unit/utils/job_runner/group_sync_job/GroupSyncJob_test.ts

```diff
@@ -0,0 +1,389 @@
+      identifier: TestUtils.generateFakeClosedGroupV2PkStr(),
+    });
+
+    const func = async () => job.run();
+    // Note: the run() function should never throw, at most it should return "permanent failure"
+    await expect(func()).to.be.not.eventually.rejected;
+  });
+
+  it('permanent failure if group is not a 03 one', async () => {
+    const job = new GroupSync.GroupSyncJob({
+      identifier: TestUtils.generateFakeClosedGroupV2PkStr().slice(2),
+    });
+    const result = await job.run();
+    expect(result).to.be.eq(RunJobResult.PermanentFailure);
+  });
+
+  it('permanent failure if user has no ed keypair', async () => {
+    Sinon.stub(UserUtils, 'getOurPubKeyStrFromCache').returns(TestUtils.generateFakePubKeyStr());
+    Sinon.stub(UserUtils, 'getUserED25519KeyPairBytes').resolves(undefined);
+    Sinon.stub(ConvoHub.use(), 'get').resolves({}); // anything not falsy
+    const job = new GroupSync.GroupSyncJob({
+      identifier: TestUtils.generateFakeClosedGroupV2PkStr(),
+    });
+    const result = await job.run();
+    expect(result).to.be.eq(RunJobResult.PermanentFailure);
+  });
+
+  it('permanent failure if user has no own conversation', async () => {
+    Sinon.stub(UserUtils, 'getOurPubKeyStrFromCache').returns(TestUtils.generateFakePubKeyStr());
+    Sinon.stub(UserUtils, 'getUserED25519KeyPairBytes').resolves({} as any); // anything not falsy
+    Sinon.stub(ConvoHub.use(), 'get').returns(undefined as any);
+    const job = new GroupSync.GroupSyncJob({
+      identifier: TestUtils.generateFakeClosedGroupV2PkStr(),
+    });
+    const result = await job.run();
+    expect(result).to.be.eq(RunJobResult.PermanentFailure);
+  });
+
+  it('calls pushChangesToGroupSwarmIfNeeded if preconditions are fine', async () => {
+    Sinon.stub(UserUtils, 'getOurPubKeyStrFromCache').returns(TestUtils.generateFakePubKeyStr());
+    Sinon.stub(UserUtils, 'getUserED25519KeyPairBytes').resolves({} as any); // anything not falsy
+    const taskedRun = Sinon.stub(GroupSync, 'pushChangesToGroupSwarmIfNeeded').resolves(
+      RunJobResult.Success
+    );
+    Sinon.stub(ConvoHub.use(), 'get').returns({} as any); // anything not falsy
+    const job = new GroupSync.GroupSyncJob({
+      identifier: TestUtils.generateFakeClosedGroupV2PkStr(),
+    });
+    const result = await job.run();
+    expect(result).to.be.eq(RunJobResult.Success);
+    expect(taskedRun.callCount).to.be.eq(1);
+  });
+});
+
+describe('GroupSyncJob resultsToSuccessfulChange', () => {
+  let sodium: LibSodiumWrappers;
+  beforeEach(async () => {
+    sodium = await getSodiumNode();
+  });
+  it('no or empty results return empty array', () => {
+    expect(
+      LibSessionUtil.batchResultsToGroupSuccessfulChange(null, {
+        allOldHashes: new Set(),
+        messages: [],
+      })
+    ).to.be.deep.eq([]);
+
+    expect(
+      LibSessionUtil.batchResultsToGroupSuccessfulChange([] as any as NotEmptyArrayOfBatchResults, {
```

Comment:
Should as any not be as unknown then as NotEmptyArrayOfBatchResults?

Reply:
updated